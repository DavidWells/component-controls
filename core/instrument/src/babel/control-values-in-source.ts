import {
  Story,
  StoryArguments,
  StoryArgument,
} from '@component-controls/specification';
import { adjustSourceLocation } from './utils';

const findArguments = (
  args: StoryArguments,
  paramName: string,
): StoryArgument | undefined => {
  let result: StoryArgument | undefined;
  for (let i = 0; i < args.length; i += 1) {
    const p = args[i];
    if (Array.isArray(p.value)) {
      result = findArguments(p.value, paramName);
      if (result) {
        break;
      }
    } else {
      if (p.value === paramName) {
        result = p;
        break;
      }
    }
  }
  return result;
};
export const addArgumentUsage = (
  story: Story,
  args: StoryArguments,
  node: any,
) => {
  const param = findArguments(args, node.name);

  if (param) {
    if (param.usage === undefined) {
      param.usage = [];
    }
    param.usage.push(adjustSourceLocation(story, node.loc));
  }
};
export const extractArgumentsUsage = (story: Story, args: StoryArguments) => {
  return {
    Identifier: (path: any) => {
      addArgumentUsage(story, args, path.node);
    },
  };
};
