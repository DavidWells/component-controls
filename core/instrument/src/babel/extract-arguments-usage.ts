import {
  Story,
  StoryArguments,
  StoryArgument,
  ArgUsageLocation,
} from '@component-controls/specification';
import { adjustSourceLocation } from './utils';

const findArguments = (
  args: StoryArguments,
  paramName: string,
): StoryArgument | undefined => {
  let result: StoryArgument | undefined;
  for (let i = 0; i < args.length; i += 1) {
    const p = args[i];
    if (Array.isArray(p.value)) {
      result = findArguments(p.value, paramName);
      if (result) {
        break;
      }
    } else {
      if (p.value === paramName) {
        result = p;
        break;
      }
    }
  }
  return result;
};
export const addArgumentUsage = (
  story: Story,
  args: StoryArguments,
  node: any,
): ArgUsageLocation | undefined => {
  const param = findArguments(args, node.name);

  if (param) {
    if (param.usage === undefined) {
      param.usage = [];
    }
    const loc = adjustSourceLocation(story, node.loc);
    const existing = param.usage.find(
      p =>
        p.loc.start.line === loc.start.line &&
        p.loc.start.column === loc.start.column &&
        p.loc.end.line === loc.end.line &&
        p.loc.end.column === loc.end.column,
    );
    const usage: ArgUsageLocation = { loc };
    if (!existing) {
      param.usage.push(usage);
      return usage;
    }
  }
  return undefined;
};
export const extractArgumentsUsage = (story: Story, args: StoryArguments) => {
  return {
    TemplateLiteral: (path: any) => {
      path.node.expressions.forEach((expression: any) => {
        if (expression.type === 'Identifier') {
          addArgumentUsage(story, args, expression);
        }
      });
    },
    JSXSpreadAttribute: (path: any) => {
      if (path.node.argument.type === 'Identifier') {
        addArgumentUsage(story, args, path.node.argument);
      }
    },
    Identifier: (path: any) => {
      addArgumentUsage(story, args, path.node);
    },
    Property: (path: any) => {
      const node = path.node;
      if (node.value.type === 'Identifier' && node.key.type === 'Identifier') {
        const usage = addArgumentUsage(story, args, node.value);
        if (usage) {
          usage.name = {
            loc: adjustSourceLocation(story, node.key.loc),
            name: node.key.name,
          };
          path.skip();
        }
      }
    },
  };
};
