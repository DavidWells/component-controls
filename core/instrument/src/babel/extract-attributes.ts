import { StoryParameters } from '@component-controls/specification';

interface StoryAttribute {
  name: string;
  value: any;
}
const nodeToValue = (node: any): any => {
  if (node) {
    switch (node.type) {
      case 'BooleanLiteral':
      case 'NumericLiteral':
      case 'StringLiteral':
        return node.value;
      case 'Identifier':
        return node.name;
      case 'Property':
        return node.name;
      case 'ObjectProperty':
        console.log('ObjectProperty', node.key.value);
        return node.key.value;

      case 'Literal':
        return node.raw;
      case 'RegExpLiteral':
        const value = node.raw ?? node.extra ? node.extra.raw : undefined;
        // remove leading trailing slashes for string to reg xonversion
        return typeof value === 'string'
          ? value.replace(/^\/|\/$/g, '')
          : value;
      case 'MemberExpression':
        return `${node.object.name}.${node.property.name}`;
      case 'ObjectExpression':
        return extractAttributes(node);
      case 'ArrayExpression':
        if (Array.isArray(node.elements)) {
          return node.elements.map((v: any) => nodeToValue(v));
        }
        break;
      default:
        // console.log(property.value);
        return undefined;
    }
  }
  return undefined;
};
const nodeToAttribute = (node: any): StoryAttribute | undefined => {
  const value = node.value || node;
  const name = node.key ? node.key.name ?? node.key.value : node.name;
  const retVal = nodeToValue(value);
  return retVal !== undefined
    ? name
      ? { value: retVal, name }
      : retVal
    : undefined;
};
export const extractAttributes = (
  node: any,
): StoryParameters | any | undefined => {
  if (node) {
    if (node.properties) {
      const attributes: StoryParameters = node.properties.reduce(
        (acc: StoryParameters, propNode: any) => {
          const attribute = nodeToAttribute(propNode);
          if (attribute) {
            return { ...acc, [attribute.name]: attribute.value };
          } else {
            return acc;
          }
        },
        {},
      );
      return attributes;
    }
    const attribute = nodeToAttribute(node);
    if (attribute) {
      return attribute.value;
    }
  }
  return undefined;
};
