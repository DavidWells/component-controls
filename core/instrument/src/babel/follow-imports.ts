import * as fs from 'fs';
import * as resolve from 'resolve';
import * as parser from '@babel/parser';
import * as path from 'path';
import { File } from '@babel/types';
import traverse from '@babel/traverse';
import { CodeLocation } from '@component-controls/specification';
import { ImportTypes, traverseImports } from './extract-imports';
import {
  traverseExports,
  ExportTypes,
  ExportType,
  EXPORT_ALL,
} from './extract-exports';

export interface FollowImportType {
  exportedAs: string;
  from: string;
  filePath?: string;
  loc?: CodeLocation;
  source?: string;
}

export const followImports = (
  importName: string,
  filePath: string,
  parserOptions?: parser.ParserOptions,
  resolveOptions?: resolve.SyncOpts,
  initialAST?: File,
): FollowImportType | undefined => {
  const source = fs.readFileSync(filePath, 'utf8');
  const ast = initialAST || parser.parse(source, parserOptions);

  const baseImportedName = importName.split('.')[0];

  const exports: ExportTypes = {
    named: {},
  };
  traverse(ast, traverseExports(exports));
  const folderName = path.dirname(filePath);
  const findExport =
    baseImportedName === 'default' || baseImportedName === 'namespace'
      ? exports.default
      : exports.named[baseImportedName];
  if (findExport !== undefined) {
    if (!findExport.from) {
      return {
        filePath,
        exportedAs: findExport.name,
        loc: findExport.loc,
        from: '',
        source,
      };
    } else {
      const resolvedFilePath = resolve.sync(findExport.from, {
        ...resolveOptions,
        basedir: folderName,
      });
      const imported = followImports(
        findExport.internalName,
        resolvedFilePath,
        parserOptions,
        resolveOptions,
      );
      return imported ? { ...imported, from: findExport.from } : undefined;
    }
  }
  const allExports: ExportType[] = Object.keys(exports.named).reduce(
    (acc: ExportType[], key: string) => [...acc, exports.named[key]],
    [],
  );
  let foundInExportAll: FollowImportType | undefined;
  allExports
    .filter(e => e.internalName === EXPORT_ALL && e.from)
    .some(e => {
      if (e.from) {
        const resolvedFilePath = resolve.sync(e.from, {
          ...resolveOptions,
          basedir: folderName,
        });
        foundInExportAll = followImports(
          baseImportedName,
          resolvedFilePath,
          parserOptions,
          resolveOptions,
        );
      }
      return { ...foundInExportAll, from: e.from };
    });
  if (foundInExportAll) {
    return foundInExportAll;
  }
  const imports: ImportTypes = {};
  traverse(ast, traverseImports(imports));
  const findImport = imports[baseImportedName];
  if (findImport) {
    try {
      const resolvedFilePath = resolve.sync(findImport.from, {
        ...resolveOptions,
        basedir: folderName,
      });
      const imported = followImports(
        findImport.importedName,
        resolvedFilePath,
        parserOptions,
        resolveOptions,
      );
      return imported ? { ...imported, from: findImport.from } : undefined;
    } catch (e) {
      //non-exxsiting file
      return { exportedAs: findImport.importedName, from: findImport.from };
    }
  }
  return undefined;
};
