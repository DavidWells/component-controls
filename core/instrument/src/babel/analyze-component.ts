import fs from 'fs';
import * as parser from '@babel/parser';
import traverse, { TraverseOptions } from '@babel/traverse';
import { Component, JSXNode, JSXTree } from '@component-controls/core';
import { collectAttributes } from './extract-attributes';
import { InstrumentOptions } from '../types';

type JSXLinkedNode = JSXNode & { parent?: JSXNode };
type JSXLinkedTree = JSXLinkedNode[];

export const traverseJSX = (jsx: JSXLinkedTree): TraverseOptions => {
  let current: JSXLinkedNode = { children: jsx };
  return {
    JSXElement: {
      enter(path) {
        const node = path.node;
        const name = node.openingElement.name as { name: string };
        const attributes = collectAttributes(node.openingElement);
        const jsxNode: JSXLinkedNode = {
          name: name.name,
          parent: current,
          children: [],
          attributes: Object.keys(attributes),
        };
        if (current.children) {
          current.children.push(jsxNode);
        }
        current = jsxNode;
      },

      exit() {
        current = current.parent as JSXLinkedNode;
      },
    },
  };
};
export const analyze_components = (
  component: Component,
  filePath: string,
  options?: InstrumentOptions,
): void => {
  const { parser: parserOptions } = options || {};
  const source = fs.readFileSync(filePath, 'utf8');
  const ast = parser.parse(source, parserOptions);
  const jsx: JSXLinkedTree = [];
  traverse(ast, traverseJSX(jsx));
  const mapJSXTree = (input?: JSXLinkedTree): JSXTree => {
    return input
      ? // eslint-disable-next-line @typescript-eslint/no-unused-vars
        input.map(({ parent, children, ...rest }) => ({
          children: mapJSXTree(children),
          ...rest,
        }))
      : [];
  };
  component.jsx = mapJSXTree(jsx);
};
