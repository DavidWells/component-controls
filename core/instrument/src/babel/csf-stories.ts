import {
  StoriesStore,
  Story,
  StoriesDoc,
  Stories,
} from '@component-controls/specification';
import { File } from '@babel/types';
import traverse from '@babel/traverse';
import { extractFunctionParameters } from './extract-function-parameters';
import { extractAttributes } from './extract-attributes';
import { followStoryImport } from './follow-imports';
import { componentsFromParams } from '../misc/component-attributes';
import { sourceLocation } from '../misc/source-location';
import { ParseStorieReturnType, InstrumentOptions } from '../types';

export const extractCSFStories = (
  ast: File,
  _options: InstrumentOptions,
  { source, filePath }: { source: string; filePath: string },
): ParseStorieReturnType => {
  const globals: Stories = {};
  const localStories: Stories = {};

  const extractArrowFunction = (
    path: any,
    declaration: any,
  ): Story | undefined => {
    if (declaration.init) {
      switch (declaration.init.type) {
        case 'ArrowFunctionExpression': {
          const el = declaration.init.body;
          const name = declaration.id.name;
          const story: Story = {
            loc: sourceLocation(el.loc),
            name,
            id: name,
          };
          traverse(path.node, extractFunctionParameters(story), path.scope);
          return story;
        }
        case 'Identifier': {
          return followStoryImport(
            declaration.id.name,
            declaration.init.name,
            filePath,
            source,
            _options,
            ast,
          );
        }
      }
    }
    return undefined;
  };
  let components: { [key: string]: string | undefined } = {};
  const store: StoriesStore = {
    stories: {},
    docs: {},
    components: {},
    packages: {},
  };
  traverse(ast as any, {
    ExportDefaultDeclaration: (path: any) => {
      const { declaration } = path.node;
      const attributes = extractAttributes(declaration);

      const { title } = attributes || {};
      if (typeof title === 'string') {
        const attrComponents = componentsFromParams(attributes);
        components = attrComponents.reduce(
          (acc, componentName) => ({ ...acc, [componentName]: undefined }),
          components,
        );
        const doc: StoriesDoc = {
          ...attributes,
          components: {},
        };
        if (attrComponents.length > 0) {
          doc.component = attrComponents[0];
        }
        store.docs[title] = doc;
      }
    },
    AssignmentExpression: (path: any) => {
      const node = path.node;
      if (
        node.left.type === 'MemberExpression' &&
        node.left.property.type === 'Identifier' &&
        node.left.property.name === 'story' &&
        node.right.type === 'ObjectExpression'
      ) {
        const storyName = node.left.object.name;
        const { name = storyName, ...attributes } = extractAttributes(
          node.right,
        );
        globals[storyName] = {
          ...attributes,
          name,
        };

        if (store.stories[storyName]) {
          const attrComponents = componentsFromParams(attributes);
          components = attrComponents.reduce(
            (acc, componentName) => ({ ...acc, [componentName]: undefined }),
            components,
          );
          const story: Story = {
            ...attributes,
            ...store.stories[storyName],
            name,
          };
          if (attrComponents.length > 0) {
            story.component = attrComponents[0];
          }
          store.stories[storyName] = story;
        }
      }
    },
    VariableDeclaration: (path: any) => {
      const { declarations } = path.node;

      if (Array.isArray(declarations) && declarations.length > 0) {
        const declaration = declarations[0];
        if (declaration) {
          const name = declaration.id.name;
          //check if it was a named export
          if (!store.stories[name]) {
            const story = extractArrowFunction(path, declaration);
            if (story && story.name) {
              localStories[story.name] = story;
            }
          }
        }
      }
    },
    ExportSpecifier: (path: any) => {
      const { node } = path;
      const localName = node.local.name;
      const exportedName = node.exported.name;
      const story = localStories[localName];
      if (story) {
        const global = globals[localName];
        if (global) {
          localStories[localName] = {
            ...story,
            ...global,
          };
        }
        store.stories[exportedName] = story;
      }
    },
    ExportNamedDeclaration: (path: any) => {
      const { declaration } = path.node;
      if (declaration) {
        const { declarations } = declaration;
        if (Array.isArray(declarations) && declarations.length > 0) {
          const story = extractArrowFunction(path, declarations[0]);
          if (story) {
            const name = story.name;
            store.stories[name] = {
              ...story,
              ...globals[name],
            };
          }
        }
      }
    },
  });
  if (Object.keys(store.docs).length === 1) {
    //@ts-ignore
    store.docs[Object.keys(store.docs)[0]].components = components;
  } else {
    throw new Error(`stories should have one default export`);
  }
  return store;
};
