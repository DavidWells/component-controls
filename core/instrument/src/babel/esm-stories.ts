import { Story, Document, Stories } from '@component-controls/core';
import { File } from '@babel/types';
import traverse from '@babel/traverse';
import { extractFunctionParameters } from './extract-function-parameters';
import { extractAttributes } from './extract-attributes';
import { followStoryImport } from './follow-imports';
import { componentsFromParams } from '../misc/component-attributes';
import { sourceLocation } from '../misc/source-location';
import {
  LoadingDocStore,
  ParseStorieReturnType,
  InstrumentOptions,
} from '../types';

export const extractCSFStories = (
  ast: File,
  _options: InstrumentOptions,
  { source, filePath }: { source: string; filePath: string },
): ParseStorieReturnType => {
  const globals: Stories = {};
  const localFunctions: Stories = {};

  const extractFunction = (path: any, declaration: any): Story | undefined => {
    if (declaration.init) {
      switch (declaration.init.type) {
        case 'ArrowFunctionExpression': {
          const el = declaration.init.body;
          const name = declaration.id.name;
          const story: Story = {
            loc: sourceLocation(el.loc),
            name,
            id: name,
          };
          traverse(path.node, extractFunctionParameters(story), path.scope);
          return story;
        }
        case 'Identifier': {
          return followStoryImport(
            declaration.id.name,
            declaration.init.name,
            filePath,
            source,
            _options,
            ast,
          );
        }
        case 'CallExpression': {
          //template.bind
          if (declaration.init.callee?.property?.name === 'bind') {
            const callee = declaration.init.callee?.object;
            if (callee?.name) {
              const template = localFunctions[callee.name];
              if (template) {
                const name = declaration.id.name;
                const story: Story = {
                  loc: template.loc,
                  name,
                  id: name,
                  arguments: template.arguments,
                };
                return story;
              }
            }
          }
        }
      }
    } else if (declaration.type === 'FunctionDeclaration') {
      const name = declaration.id.name;
      const story: Story = {
        loc: sourceLocation(declaration.body.loc),
        name,
        id: name,
      };
      traverse(path.node, extractFunctionParameters(story), path.scope);
      return story;
    }
    return undefined;
  };
  let components: { [key: string]: string | undefined } = {};
  const store: LoadingDocStore = {
    stories: {},
    doc: undefined,
    components: {},
    packages: {},
  };
  traverse(ast as any, {
    ExportDefaultDeclaration: (path: any) => {
      const { declaration } = path.node;
      const attributes = extractAttributes(declaration);

      const { title: docTitle, name } = attributes || {};
      const title = docTitle || name;
      if (typeof title === 'string') {
        const attrComponents = componentsFromParams(attributes);
        components = attrComponents.reduce(
          (acc, componentName) => ({ ...acc, [componentName]: undefined }),
          components,
        );
        const doc: Document = {
          ...attributes,
          title,
          components: {},
        };
        if (attrComponents.length > 0) {
          doc.component = attrComponents[0];
        }

        store.doc = doc;
      }
    },
    AssignmentExpression: (path: any) => {
      const node = path.node;
      const storyExport = node.left?.object?.name;
      if (
        node.left.type === 'MemberExpression' &&
        node.left.property.type === 'Identifier' &&
        store.stories[storyExport]
      ) {
        const extractedValue = extractAttributes(node.right);
        const extractedProps =
          node.left.property.name === 'story'
            ? extractedValue
            : { [node.left.property.name]: extractedValue };
        const { name, storyName, ...attributes } = extractedProps;

        const nameAttr = storyName || name;
        if (store.stories[storyExport]) {
          const attrComponents = componentsFromParams(attributes);
          components = attrComponents.reduce(
            (acc, componentName) => ({ ...acc, [componentName]: undefined }),
            components,
          );
          const story: Story = {
            ...attributes,
            ...store.stories[storyExport],
          };
          if (nameAttr) {
            story.name = nameAttr;
          }
          store.stories[storyExport] = story;
          globals[storyExport] = story;
        }
      }
    },
    VariableDeclaration: (path: any) => {
      const { declarations } = path.node;

      if (Array.isArray(declarations) && declarations.length > 0) {
        const declaration = declarations[0];
        if (declaration) {
          const name = declaration.id.name;
          //check if it was a named export
          if (!store.stories[name]) {
            const story = extractFunction(path, declaration);
            if (story && story.name) {
              localFunctions[story.name] = story;
            }
          }
        }
      }
    },
    ExportSpecifier: (path: any) => {
      const { node } = path;
      const localName = node.local.name;
      const exportedName = node.exported.name;
      const story = localFunctions[localName];
      if (story) {
        const global = globals[localName];
        if (global) {
          localFunctions[localName] = {
            ...story,
            ...global,
          };
        }
        store.stories[exportedName] = story;
      }
    },
    ExportNamedDeclaration: (path: any) => {
      const { declaration } = path.node;
      if (declaration) {
        const { declarations } = declaration;
        if (declarations) {
          if (Array.isArray(declarations) && declarations.length > 0) {
            const story = extractFunction(path, declarations[0]);
            if (story) {
              const name = story.name;
              store.stories[name] = {
                ...story,
                ...globals[name],
              };
            }
          }
        } else {
          if (declaration.type === 'FunctionDeclaration') {
            const story = extractFunction(path, declaration);
            if (story) {
              const name = story.name;
              store.stories[name] = {
                ...story,
                ...globals[name],
              };
            }
          }
        }
      }
    },
  });
  if (store.doc) {
    store.doc.componentsLookup = components as Document['componentsLookup'];
  } else {
    throw new Error(`esm files should have one default export`);
  }
  return store;
};
