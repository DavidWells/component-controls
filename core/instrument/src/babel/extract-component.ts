import path from 'path';
import { File } from '@babel/types';
import * as resolve from 'resolve';
import {
  Component,
  Document,
  PackageInfo,
  Imports,
  ImportTypes,
} from '@component-controls/core';
import { componentKey } from '../misc/hashStore';
import { followImports } from './follow-imports';
import { analyze_components } from './analyze-component';
import { packageInfo } from '../misc/package-info';
import { readSourceFile } from '../misc/source-options';

import { LoadingDocStore, InstrumentOptions } from '../types';

interface ComponentParseData {
  component?: Component;
  componentPackage?: PackageInfo;
}
const globalCache: {
  [filePath: string]: ComponentParseData;
} = {};
export const extractComponent = async (
  componentName: string,
  filePath: string,
  source?: string,
  options?: InstrumentOptions,
  initialAST?: File,
): Promise<ComponentParseData> => {
  const cacheKey = `${filePath}-${componentName}`;
  if (globalCache[cacheKey]) {
    return globalCache[cacheKey];
  }
  const follow = followImports(
    componentName,
    filePath,
    source,
    options,
    initialAST,
  );
  const { components, resolver: resolveOptions } = options || {};
  let component: Component;
  let componentPackage: PackageInfo | undefined;
  if (follow) {
    component = {
      name: componentName,
    };
    if (follow.from) {
      component.from = follow.from;
    }
    if (follow.imports) {
      const imports: ImportTypes = follow.imports;
      const allImports = Object.keys(imports).reduce((acc: Imports, key) => {
        const { name, from, importedName } = imports[key];
        let importKey = undefined;
        if (follow.filePath && from.startsWith('.')) {
          try {
            const fileName = resolve.sync(from, {
              ...resolveOptions,
              basedir: path.dirname(follow.filePath),
            });
            const followImport = followImports(
              importedName,
              fileName,
              undefined,
              options,
              undefined,
            );
            if (followImport?.filePath) {
              importKey = componentKey(followImport.filePath, importedName);
            }
          } catch (e) {}
        }

        if (acc[from]) {
          return {
            ...acc,
            [from]: [
              ...acc[from],
              { name, from, importedName, key: importKey },
            ],
          };
        }
        return {
          ...acc,
          [from]: [{ name, from, importedName, key: importKey }],
        };
      }, {});
      component.externalDependencies = Object.keys(allImports)
        .filter(key => !key.startsWith('.'))
        .reduce(
          (acc, key) => ({ ...acc, [key]: (allImports as Imports)[key] }),
          {},
        );
      component.localDependencies = Object.keys(allImports)
        .filter(key => key.startsWith('.'))
        .reduce((acc, key) => {
          return { ...acc, [key]: (allImports as Imports)[key] };
        }, {});
    }
    if (follow.importedName) {
      component.importedName = follow.importedName;
    }
    if (follow.filePath) {
      if (follow.imports) {
        analyze_components(component, follow.filePath, options);
      }

      component.request = follow.filePath;
      if (components && typeof components.resolvePropsFile === 'function') {
        const propsFile = components.resolvePropsFile(
          componentName,
          follow.filePath,
        );
        if (propsFile !== component.request) {
          component.propsInfoFile = propsFile;
        }
      }
      const saveSource = readSourceFile(
        components?.sourceFiles,
        follow.source,
        componentName,
        follow.filePath,
      );
      if (saveSource) {
        component.source = saveSource;
        component.loc = follow.loc;
      }
    }
    componentPackage = await packageInfo(
      componentName,
      follow.originalFilePath,
      options?.components?.package,
    );
  } else {
    component = {
      name: componentName,
    };
  }

  globalCache[cacheKey] = { component, componentPackage };
  return globalCache[cacheKey];
};

export const extractStoreComponent = async (
  store: LoadingDocStore,
  filePath: string,
  source: string,
  options?: InstrumentOptions,
  initialAST?: File,
): Promise<void> => {
  if (store.doc) {
    const doc: Document = store.doc;
    if (doc.componentsLookup) {
      const componentNames = Object.keys(doc.componentsLookup);
      if (componentNames) {
        for (const componentName of componentNames) {
          const { component, componentPackage } = await extractComponent(
            componentName,
            filePath,
            source,
            options,
            initialAST,
          );
          if (component) {
            if (componentPackage) {
              store.packages[componentPackage.fileHash] = componentPackage;
              component.package = componentPackage.fileHash;
            }
            const key = componentKey(
              component.request ?? filePath,
              componentName,
            );
            store.components[key] = component;
            doc.componentsLookup[componentName] = key;
          }
        }
      }
    }
  }
};
