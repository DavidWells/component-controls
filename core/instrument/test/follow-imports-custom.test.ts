import * as path from 'path';
import {
  defaultParserOptions,
  defaultResolveOptions,
  defaultComponentOptions,
} from '../src/index';
import { followImports } from '../src/babel/follow-imports';

describe('follow-imports-custom', () => {
  const loadTestFiles = (
    componentName: string,
    fileName: string,
    source?: string,
  ) => {
    it(componentName, async () => {
      expect(
        await followImports(componentName, fileName, source, {
          parser: defaultParserOptions,
          resolver: defaultResolveOptions,
          components: defaultComponentOptions,
        }),
      ).toMatchSnapshot();
    });
  };
  loadTestFiles(
    'BooleanEditor',
    path.resolve(
      __dirname,
      '../../../ui/editors/src/BooleanEditor/BooleanEditor.stories.tsx',
    ),
    `
    import React from 'react';
    import { ControlTypes } from '@component-controls/specification';
    import { BooleanEditor } from './BooleanEditor';
    
    export default {
        title: 'Editors/BooleanEditor',
        component: BooleanEditor,
    };
    
    export const sample = () => {
        const [state, setState] = React.useState(false);
        return (
            <BooleanEditor
                name="prop"
                onChange={(name, newVal) => setState(newVal)}
                prop={{ type: ControlTypes.BOOLEAN, value: state }}
            />
        );
    };
`,
  );

  loadTestFiles(
    'Markdown',
    path.resolve(
      __dirname,
      '../../../ui/components/src/Markdown/Markdown.stories.tsx',
    ),
  );

  loadTestFiles(
    'Title',
    path.resolve(__dirname, '../../../ui/blocks/src/Title/Title.stories.tsx'),
  );
});
