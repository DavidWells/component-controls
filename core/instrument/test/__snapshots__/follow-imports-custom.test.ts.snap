// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`follow-imports-custom BooleanEditor 1`] = `[Function]`;

exports[`follow-imports-custom Markdown 1`] = `[Function]`;

exports[`follow-imports-custom Title 1`] = `[Function]`;
